lib/mix/lib/mix/compilers/erlang.ex

```diff
@@ -188,16 +190,60 @@ defmodule Mix.Compilers.Erlang do
     artifact |> Path.basename |> Path.rootname
   end
 
+  # The manifest file contains a list of {dest, warnings} tuples
   defp read_manifest(file) do
-    case File.read(file) do
-      {:ok, contents} -> String.split(contents, "\n")
-      {:error, _} -> []
+    try do
+      file |> File.read!() |> :erlang.binary_to_term()
+    rescue
+      _ -> []
+    else
+      {@manifest_vsn, data} when is_list(data) -> data
+      _ -> []
     end
   end
 
   defp write_manifest(file, entries, timestamp) do
     Path.dirname(file) |> File.mkdir_p!
-    File.write!(file, Enum.join(entries, "\n"))
+    File.write!(file, :erlang.term_to_binary({@manifest_vsn, entries}))
     File.touch!(file, timestamp)
   end
+
+  defp do_compile({input, output}, callback, timestamp, verbose) do
+    case callback.(input, output) do
+      {:ok, _, warnings} ->
+        File.touch!(output, timestamp)
+        verbose && Mix.shell.info "Compiled #{input}"
+        {:ok, [{output, warnings}], warnings, []}
+      {:error, errors, warnings} ->
+        {:error, [], warnings, errors}
+    end
+  end
+
+  defp combine_results(result1, result2) do
+    {status1, new_entries1, warnings1, errors1} = result1
+    {status2, new_entries2, warnings2, errors2} = result2
+    status = if status1 == :error or status2 == :error, do: :error, else: :ok
+    {status, new_entries1 ++ new_entries2, warnings1 ++ warnings2, errors1 ++ errors2}
+  end
+
+  defp manifest_warnings(entries) do
+    Enum.flat_map(entries, fn {_, warnings} ->
+      to_diagnostics(warnings, :warning)
+    end)
+  end
+
+  defp to_diagnostics(warnings_or_errors, severity) do
+    Enum.flat_map(warnings_or_errors, fn {file, issues} ->
```

Comment:
This one can also be written as:

    for {file, issues} <- warnings_or_errors,
        {line, module, data} <- issues do
      ...